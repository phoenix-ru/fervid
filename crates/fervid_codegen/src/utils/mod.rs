// to_ident_or_str

use std::fmt::{Write, Error};

use fervid_core::{FervidAtom, StrOrExpr};
use swc_core::{common::Span, ecma::ast::{ComputedPropName, Ident, EsReserved, IdentName, PropName, Str}};

/// Adapted from SWC Ident::verify_symbol
#[inline]
pub fn is_valid_ident(s: &str) -> bool {
    if s.is_reserved() || s.is_reserved_in_strict_mode(true) || s.is_reserved_in_strict_bind() {
        return false;
    }

    let mut chars = s.chars();

    if let Some(first) = chars.next() {
        if Ident::is_valid_start(first) && chars.all(Ident::is_valid_continue) {
            return true;
        }
    }

    false
}

pub fn str_to_propname(s: &str, span: Span) -> PropName {
    if is_valid_ident(s) {
        PropName::Ident(IdentName { span, sym: s.into() })
    } else {
        PropName::Str(Str {
            span,
            value: s.into(),
            raw: None,
        })
    }
}

pub fn atom_to_propname(sym: FervidAtom, span: Span) -> PropName {
    if is_valid_ident(&sym) {
        PropName::Ident(IdentName { span, sym })
    } else {
        PropName::Str(Str {
            span,
            value: sym,
            raw: None,
        })
    }
}

pub fn str_or_expr_to_propname(str_or_expr: StrOrExpr, span: Span) -> PropName {
    match str_or_expr {
        StrOrExpr::Str(sym) => atom_to_propname(sym, span),
        StrOrExpr::Expr(expr) => PropName::Computed(ComputedPropName {
            span,
            expr,
        }),
    }
}

pub fn to_camelcase(s: &str, buf: &mut impl Write) -> Result<(), Error> {
    for (idx, word) in s.split('-').enumerate() {
        if idx == 0 {
            buf.write_str(word)?;
            continue;
        }

        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                buf.write_char(ch_component)?;
            }

            // Push the rest of the word
            buf.write_str(&word[ch.len_utf8()..])?;
        }
    }

    Ok(())
}
