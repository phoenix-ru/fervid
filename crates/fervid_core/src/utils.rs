use swc_core::{common::Span, ecma::ast::{ComputedPropName, EsReserved, Ident, IdentName, PropName, Str}};

use crate::{AttributeOrBinding, FervidAtom, StrOrExpr, VBindDirective};

/// Checks whether the attributes name is the same as `expected_name`
#[inline]
pub fn check_attribute_name(attr: &AttributeOrBinding, expected_name: &str) -> bool {
    matches!(attr,
        AttributeOrBinding::RegularAttribute { name, .. } |
        AttributeOrBinding::VBind(VBindDirective { argument: Some(StrOrExpr::Str(name)), .. })
        if name == expected_name
    )
}

/// Adapted from SWC Ident::verify_symbol
#[inline]
pub fn is_valid_ident(s: &str) -> bool {
    if s.is_reserved() || s.is_reserved_in_strict_mode(true) || s.is_reserved_in_strict_bind() {
        return false;
    }

    is_valid_propname(s)
}

#[inline]
pub fn is_valid_propname(s: &str) -> bool {
    let mut chars = s.chars();

    if let Some(first) = chars.next() {
        if Ident::is_valid_start(first) && chars.all(Ident::is_valid_continue) {
            return true;
        }
    }

    false
}

pub fn str_to_propname(s: &str, span: Span) -> PropName {
    if is_valid_propname(s) {
        PropName::Ident(IdentName { span, sym: s.into() })
    } else {
        PropName::Str(Str {
            span,
            value: s.into(),
            raw: None,
        })
    }
}

pub fn atom_to_propname(sym: FervidAtom, span: Span) -> PropName {
    if is_valid_propname(&sym) {
        PropName::Ident(IdentName { span, sym })
    } else {
        PropName::Str(Str {
            span,
            value: sym,
            raw: None,
        })
    }
}

pub fn str_or_expr_to_propname(str_or_expr: StrOrExpr, span: Span) -> PropName {
    match str_or_expr {
        StrOrExpr::Str(sym) => atom_to_propname(sym, span),
        StrOrExpr::Expr(expr) => PropName::Computed(ComputedPropName {
            span,
            expr,
        }),
    }
}
