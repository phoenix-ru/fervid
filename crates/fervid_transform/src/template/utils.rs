use fervid_core::FervidAtom;
use swc_core::{common::DUMMY_SP, ecma::ast::{ArrowExpr, BindingIdent, BlockStmtOrExpr, Expr, Ident, Pat}};

/// `foo-bar-baz` -> `FooBarBaz`
#[inline]
pub(crate) fn to_pascal_case(raw: &str, out: &mut String) {
    for word in raw.split('-') {
        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                out.push(ch_component);
            }

            // Push the rest of the word
            out.push_str(&word[ch.len_utf8()..]);
        }
    }
}

/// `foo-bar-baz` -> `fooBarBaz`
#[inline]
pub(crate) fn to_camel_case(raw: &str, out: &mut String) {
    for (idx, word) in raw.split('-').enumerate() {
        if idx == 0 {
            out.push_str(word);
            continue;
        }

        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                out.push(ch_component);
            }

            // Push the rest of the word
            out.push_str(&word[ch.len_utf8()..]);
        }
    }
}

/// Wraps `expr` to `$event => (expr)`
#[inline]
pub fn wrap_in_event_arrow(expr: Box<Expr>) -> Box<Expr> {
    let evt_param = Pat::Ident(BindingIdent {
        id: Ident {
            span: DUMMY_SP,
            ctxt: Default::default(),
            sym: FervidAtom::from("$event"),
            optional: false,
        },
        type_ann: None,
    });

    Box::new(Expr::Arrow(ArrowExpr {
        span: DUMMY_SP,
        ctxt: Default::default(),
        params: vec![evt_param],
        body: Box::new(BlockStmtOrExpr::Expr(expr)),
        is_async: false,
        is_generator: false,
        type_params: None,
        return_type: None,
    }))
}
