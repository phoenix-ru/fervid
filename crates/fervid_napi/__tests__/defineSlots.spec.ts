import { describe, expect, test } from 'vitest'
import { assertCode, compile } from './utils'

describe('defineSlots()', () => {
  test('basic usage', () => {
    const { content } = compile(`
      <script setup lang="ts">
      const slots = defineSlots<{
        default: { msg: string }
      }>()
      </script>
    `)
    assertCode(content)
    expect(content).toMatch(`const slots = _useSlots()`)
    expect(content).not.toMatch('defineSlots')
  })

  test('w/o return value', () => {
    const { content } = compile(`
      <script setup lang="ts">
      defineSlots<{
        default: { msg: string }
      }>()
      </script>
    `)
    assertCode(content)
    expect(content).not.toMatch('defineSlots')
    expect(content).not.toMatch(`_useSlots`)
  })

  test('w/o generic params', () => {
    const { content } = compile(`
      <script setup>
      const slots = defineSlots()
      </script>
    `)
    assertCode(content)
    expect(content).toMatch(`const slots = _useSlots()`)
    expect(content).not.toMatch('defineSlots')
  })

  test('error on duplicate', () => {
    const { errors } = compile(`
      <script setup>
      defineSlots()
      defineSlots()
      </script>
    `)
    expect(errors.length).toBe(1)
    expect(errors[0].message).toMatch('DuplicateDefineSlots')
  })

  test('error on arguments', () => {
    const { errors } = compile(`
      <script setup>
      const slots = defineSlots({
        default: () => {}
      })
      </script>
    `)
    expect(errors.length).toBe(1)
    expect(errors[0].message).toMatch('DefineSlotsArguments')
  })
})
