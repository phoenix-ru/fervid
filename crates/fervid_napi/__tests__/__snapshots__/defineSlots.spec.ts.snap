// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineSlots() > basic usage 1`] = `
"import { defineComponent as _defineComponent, useSlots as _useSlots } from "vue";
export default _defineComponent({
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const slots = _useSlots();
        return {
            slots
        };
    }
});
"
`;

exports[`defineSlots() > w/o generic params 1`] = `
"import { useSlots as _useSlots } from "vue";
export default {
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const slots = _useSlots();
        return {
            slots
        };
    }
};
"
`;

exports[`defineSlots() > w/o return value 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
    }
});
"
`;
